web/shared/src/typing_status.ts

```diff
@@ -1,26 +1,64 @@
 import _ from "lodash";
 import assert from "minimalistic-assert";
 
+type StreamTopic = {
+    stream_id: number;
+    topic: string;
+};
+type Recipient = number[] | StreamTopic;
 type TypingStatusWorker = {
     get_current_time: () => number;
-    notify_server_start: (recipient_ids: number[]) => void;
-    notify_server_stop: (recipient_ids: number[]) => void;
+    notify_server_start: (recipient: Recipient) => void;
+    notify_server_stop: (recipient: Recipient) => void;
 };
 
 type TypingStatusState = {
-    current_recipient_ids: number[];
+    current_recipient: Recipient;
     next_send_start_time: number;
     idle_timer: ReturnType<typeof setTimeout>;
 };
 
+function message_type(recipient: Recipient): "direct" | "stream" {
+    if (Array.isArray(recipient)) {
+        return "direct";
+    }
+    return "stream";
+}
+
+function lower_same(a: string, b: string): boolean {
+    return a.toLowerCase() === b.toLowerCase();
+}
+
+function same_stream_and_topic(a: StreamTopic, b: StreamTopic): boolean {
+    // Streams and topics are case-insensitive.
+    return a.stream_id === b.stream_id && lower_same(a.topic, b.topic);
+}
+
+function same_recipient(a: Recipient | null, b: Recipient | null): boolean {
+    if (a === null || b === null) {
+        return false;
+    }
+
+    if (message_type(a) === "direct" && message_type(b) === "direct") {
+        return _.isEqual(a, b);
+    } else if (message_type(a) === "stream" && message_type(b) === "stream") {
+        // type assertions to avoid linter error
+        const aStreamTopic = a as StreamTopic;
+        const bStreamTopic = b as StreamTopic;
```

Comment:
These can likely be avoided by just asserting they're not integers, given the types.